src/graders/test-diversity.arr

```diff
@@ -0,0 +1,664 @@
+  student = cases (DiversityGuardBlock) reason:
+  | parser-error(_) =>
+    "Cannot find your function definition because we cannot parse your file."
+  | fn-not-defined(name) =>
+    "Cannot find a function named " + MD.escape-inline-code(name)
+    + " in your code."
+  | run-error(_) =>
+    "We ran into an error trying to run your function and its tests."
+  | invalid-result(_) =>
+    "The autograder received an unexpected response running your tests. "
+    + "Please report this bug to a staff member."
+  | too-few-inputs(name, expected, actual) =>
+    "Your tests for function " + MD.escape-inline-code(name)
+    + " called it with " + num-to-string(actual) + " distinct sets of arguments. "
+    + " A good test suite should make at least " + num-to-string(expected)
+    + " distinct calls."
+  | too-few-outputs(name, expected, actual) =>
+     "Your tests for function " + MD.escape-inline-code(name)
+     + " caused it to produce " + num-to-string(actual) + " different outputs."
+     + " A good test suite should cause the function to produce at least "
+     + num-to-string(expected) + " different outputs.\n"
+     + "Note: This counts the _actual_ outputs your function returned, not "
+     + "the expected outcomes in your tests. This may be flagging an incorrect "
+     + "implementation _or_ an insufficient test suite."
+  end ^ G.output-markdown
+  staff = cases (DiversityGuardBlock) reason:
+  | parser-error(_) => none
+  | fn-not-defined  => none
+  | run-error(err) =>
+    ("Got the following error when trying to run wrapped file:\n```\n"
+    + MD.escape-code-block(to-repr(err))
+    + "\n```")
+    ^ G.output-markdown
+    ^ some
+  | invalid-result(raw) =>
+    json-string = raw.serialize()
+    ("Got invalid JSON response when running wrapped file:\n```\n"
+    + MD.escape-code-block(json-string)
+    + "\n```")
+    ^ G.output-markdown
+    ^ some
+  | too-few-inputs(name, expected, actual) => none
+  | too-few-outputs(name, expected, actual) => none
+  end
+  {student; staff}
+end
+
+fun mk-test-diversity-guard(
+  id :: Id,
+  deps :: List<Id>,
+  path :: String,
+  fn :: String,
+  min-in :: Number,
+  min-out :: Number
+):
+  name = "Test suite diversity for " + fn
+  checker = lam(): check-test-diversity(path, fn, min-in, min-out) end
+  GB.mk-guard(id, deps, checker, name, fmt-test-diversity)
+end
+
+# --- Name generation ---
+
+fun dummy-loc() -> SL.Srcloc block:
+  doc: ```
+  The Pyret compiler assumes that all `check` blocks have real
+  (i.e., not builtin) source locations. The well-formedness checker also rejects
+  the file if multiple expressions in a `s-block` are on "the same line". So the
+  easiest way around both of these is just to fake the line number every time.
+  ```
+  next-line := next-line + 1
+  SL.srcloc(dummy-file-name, next-line, 0, 0, next-line, 0, 0)
+end
+
+fun input-set-name(fn :: String) -> String:
+  "autograder$" + fn + "-diversity-inputs"
```

Comment:
Sorry if I was the one who told you to format it this way, this isn't guaranteed to be hygienic. Pyret replaces all `-` with `$` at [anf-loop-compiler.arr:135](https://github.com/brownplt/pyret-lang/blob/horizon/src/arr/compiler/anf-loop-compiler.arr#L135). Since Pyret disallows `-` at the beginning of an identifier, you should instead use the prefix `$autograder`.